### PR TITLE
interfaces: add gpio-memory-control interface

### PR DESCRIPTION
**Pull request by kyrofa**
Raspberry Pi systems expose GPIO-only memory at /dev/gpiomem, which provides much faster memory-mapped access than sysfs while being more constrained than /dev/mem.

Support write-access to /dev/gpiomem with a new interface named gpio-memory-control.

### CODE REVIEW

File: interfaces/builtin/gpio_memory_control.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const gpioMemoryControlConnectedPlugAppArmor = `
+# Description: Allow writing to /dev/gpiomem on kernels that provide it.
+#
+/dev/gpiomem rw,
```
Comment (zyga) [inline interfaces/builtin/gpio_memory_control.go]: Don't you need `m` as well for mmap?

**Comment by zyga** [comment]
@kyrofa once this gets a review from @jdstrand we should work with @sergiocazzolato on a spread test to make sure, you know, that it really works ;-)

File: interfaces/builtin/gpio_memory_control.go (hunk 1 above)
Comment (kyrofa) [reply interfaces/builtin/gpio_memory_control.go]: Honestly I don't know. This worked for my purposes, which given my rudimentary understanding of the lib I'm using uses mmap, but it's possible they're falling back (note that I see no denials, however).

Note also that `physical-memory-control` does not use `m` for `/dev/mem`: https://github.com/snapcore/snapd/blob/master/interfaces/builtin/physical_memory_control.go#L41 .

**Comment by kyrofa** [comment]
> [...] we should work with @sergiocazzolato on a spread test to make sure, you know, that it really works ;-)

Do you have rpis in your test rig?

File: interfaces/builtin/gpio_memory_control.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+const gpioMemoryControlBaseDeclarationSlots = `
+  gpio-memory-control:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const gpioMemoryControlConnectedPlugAppArmor = `
+# Description: Allow writing to /dev/gpiomem on kernels that provide it.
```
Comment (jdstrand) [inline interfaces/builtin/gpio_memory_control.go]: Can you update this to be:
```
# Description: Allow writing to /dev/gpiomem on kernels that provide it (eg,
# via the bcm2835-gpiomem kernel module). This allows direct access to the
# physical memory for GPIO devices (ie, a subset of /dev/mem) and therefore
# grants access to all GPIO devices on the system.
```

File: interfaces/builtin/gpio_memory_control.go (hunk 1 above)
Comment (jdstrand) [reply interfaces/builtin/gpio_memory_control.go]: 'm' is only needed when using mmap() with PROT_EXEC.

File: interfaces/builtin/gpio_memory_control.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const gpioMemoryControlSummary = `allows write access to all gpio memory`
```
Comment (jdstrand) [inline interfaces/builtin/gpio_memory_control.go]: Can you add a comment above this:
```
// https://github.com/raspberrypi/linux/blob/rpi-4.4.y/drivers/char/broadcom/bcm2835-gpiomem.c
```

**Comment by kyrofa** [comment]
Huh. Not sure what this Travis failure means means:

```
2017-11-29 20:27:08 Error executing linode:ubuntu-16.04-64:tests/unit/go : 

-----

+ mkdir -p /tmp/static-unit-tests/src/github.com/snapcore

+ cp -ar /home/gopath/src/github.com/snapcore/snapd /tmp/static-unit-tests/src/github.com/snapcore

+ chown -R test:12345 /tmp/static-unit-tests

+ rm -r /tmp/static-unit-tests/src/github.com/snapcore/snapd/vendor/github.com/ /tmp/static-unit-tests/src/github.com/snapcore/snapd/vendor/golang.org/ /tmp/static-unit-tests/src/github.com/snapcore/snapd/vendor/gopkg.in/

+ rm -rf /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/autom4te.cache /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/configure /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/test-driver /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/config.status /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/config.guess /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/config.sub /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/config.h.in /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/compile /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/install-sh /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/depcomp /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/build /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/missing /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/aclocal.m4 /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/Makefile /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/Makefile.in

+ su -l -c 'cd /tmp/static-unit-tests/src/github.com/snapcore/snapd && PATH=/home/gopath/bin:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/usr/lib/go-1.6/bin:/var/lib/snapd/snap/bin GOPATH=/tmp/static-unit-tests ./run-checks --static' test

Obtaining dependencies

Checking docs

Checking formatting

Running vet

Check for usages of http.Status*

Checking spelling errors

Checking for ineffective assignments

Checking for naked returns

Checking all interfaces have minimal spread test

fatal: Not a git repository (or any of the parent directories): .git

All good, what could possibly go wrong.

+ su -l -c 'cd /tmp/static-unit-tests/src/github.com/snapcore/snapd &&     TRAVIS_BUILD_NUMBER=18561     TRAVIS_BRANCH=master     TRAVIS_COMMIT=0235d5feb7409062e29560cff0dd7e69cf16b986     TRAVIS_JOB_NUMBER=18561.1     TRAVIS_PULL_REQUEST=4323     TRAVIS_JOB_ID=309151920     TRAVIS_REPO_SLUG=snapcore/snapd     TRAVIS_TAG=     PATH=/home/gopath/bin:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/usr/lib/go-1.6/bin:/var/lib/snapd/snap/bin     COVERMODE=     TRAVIS=true     CI=true     GOPATH=/tmp/static-unit-tests     ./run-checks --unit' test
```

File: interfaces/builtin/gpio_memory_control.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+# Description: Allow writing to /dev/gpiomem on kernels that provide it.
+#
+/dev/gpiomem rw,
+`
+
+var gpioMemoryControlConnectedPlugUDev = []string{`KERNEL=="gpiomem"`}
+
+func init() {
+	registerIface(&commonInterface{
+		name:                  "gpio-memory-control",
```
Comment (jdstrand) [inline interfaces/builtin/gpio_memory_control.go]: I'm conflicted on the name. On the one hand /dev/gpiomem gives precisely access to all GPIO memory so `gpio-memory-control` makes a lot of sense. However, there are several GPIO kernel systems and I wonder if it makes sense to have this interface reflect that this is for one, /dev/gpiomem, and therefore the interface could be named `gpiomem-control`. I think I prefer what you have, `gpio-memory-control`, but please get signoff from @niemeyer on the name.

File: tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml [hunk 5]
```diff
@@ -83,6 +83,9 @@ apps:
   fwupd:
     command: bin/run
     plugs: [ fwupd ]
+  gpio-memory-control:
+    command: bin/run
+    plugs: [ gpio-memory-control ]
```
Comment (jdstrand) [inline tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml]: \o/ missing_interface_spread_test() in run-checks did its job! :)

Comment (kyrofa) [reply tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml]: Yes, I actually said "gee, thanks" out loud!

File: interfaces/builtin/gpio_memory_control.go (hunk 3 above)
Comment (kyrofa) [reply interfaces/builtin/gpio_memory_control.go]: Definitey-- done.

File: interfaces/builtin/gpio_memory_control.go (hunk 2 above)
Comment (kyrofa) [reply interfaces/builtin/gpio_memory_control.go]: Sure thing, done.

File: interfaces/builtin/gpio_memory_control.go (hunk 4 above)
Comment (kyrofa) [reply interfaces/builtin/gpio_memory_control.go]: I am similarly conflicted, and have no strong opinions.

File: tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml (hunk 5 above)
Comment (zyga) [reply tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml]: Wow, I didn't know we had that. Thank you!!!

**Comment by mvo5** [comment]
Looks good, thanks for adding this interface! This can go in once we have sign-off for the name from @niemeyer

**Comment by zyga** [comment]
@kyrofa and you get to write a forum post about it and we get to link it via interface meta-data in snapd :)

**Pull request merged**